python/GafferUITest/GraphGadgetTest.py

```diff
@@ -1406,5 +1408,336 @@ def testChangeNodeGadgetForUnviewedNode( self ) :
+
+		# Disabling uses the passThrough instead
+		box["enabled"].setValue( False )
+		plugs, nodes = GafferUI.GraphGadget._activePlugsAndNodes( s["add2"]["sum"], c, None )
+		self.assertEqual( set( plugs ), set( [ s["add2"]["op1"], box["op1"], box["sum1"], box["BoxIn"]["__in"], box["BoxOut1"]["passThrough"], box["BoxOut1"]["__out"], box["BoxOut1"]["__switch"]["in"]["in0"] ] ) )
+		self.assertEqual( set( nodes ), set( [ box, s["add2"], s["add5"], box["BoxOut1"]["__switch"], box["BoxOut1"], box["BoxIn"] ] ) )
+
+		# If we disconnect the passThrough, nothing gets through
+		box["BoxOut1"]["passThrough"].setInput( None )
+		plugs, nodes = GafferUI.GraphGadget._activePlugsAndNodes( s["add2"]["sum"], c, None )
+		self.assertEqual( set( plugs ), set( [ s["add2"]["op1"], box["sum1"], box["BoxOut1"]["__out"], box["BoxOut1"]["__switch"]["in"]["in0"] ] ) )
+		self.assertEqual( set( nodes ), set( [ box, s["add2"], box["BoxOut1"]["__switch"], box["BoxOut1"] ] ) )
+
+
+		# Test Loop
+
+		s["loopSwitch"] = Gaffer.Switch()
+		s["loopSwitch"].setup( Gaffer.FloatPlug() )
+		for i in range( 5 ):
+			s["loopSwitchIn%i" % i ] = GafferTest.AddNode()
+			s["loopSwitchIn%i" % i ]["op1"].setValue( (i+1) * 10 ** i )
+			s["loopSwitch"]["in"][-1].setInput( s["loopSwitchIn%i" % i ]["sum"] )
+
+		s["loopSwitch"]["expr"] = Gaffer.Expression()
+		s["loopSwitch"]["expr"].setExpression( 'parent["index"] = context.get("loop:index", 4)', "python" )
+
+		s["start"] = GafferTest.AddNode()
+		s["start"]["op1"].setValue( 900000 )
+
+		s["loop"] = Gaffer.Loop()
+		s["loop"].setup( Gaffer.FloatPlug() )
+		s["loop"]["in"].setInput( s["start"]["sum"] )
+
+		s["merge"] = GafferTest.AddNode()
+		s["merge"]["op1"].setInput( s["loop"]["previous"] )
+		s["loop"]["next"].setInput( s["merge"]["sum"] )
+		s["merge"]["op2"].setInput( s["loopSwitch"]["out"] )
+
+		s["loop"]["iterations"].setValue( 4 )
+
+		self.assertEqual( s["loop"]["out"].getValue(), 904321.0 )
+
+		# We don't track through the loop for every separate loop context, we just fall back to a naive
+		# traversal that takes every input
+		plugs, nodes = GafferUI.GraphGadget._activePlugsAndNodes( s["loop"]["out"], c, None )
+		self.assertEqual( set( nodes ), set( [ s["loop"], s["start"], s["merge"], s["loopSwitch"], s["loopSwitch"]["expr"] ] + [ s["loopSwitchIn%i"%i] for i in range(5) ] ) )
+
+		# Same deal if we start from a node in the middle of the loop iteration
+		plugs, nodes = GafferUI.GraphGadget._activePlugsAndNodes( s["merge"]["sum"], c, None )
+		self.assertEqual( set( nodes ), set( [ s["loop"], s["start"], s["merge"], s["loopSwitch"], s["loopSwitch"]["expr"] ] + [ s["loopSwitchIn%i"%i] for i in range(5) ] ) )
+
+	def testActivePlugsAndNodesCancellation( self ) :
+
+		def testThreadFunc( plug, canceller ):
+			try:
+				# TODO - should probably omit canceller from this API, and just take a context?
```

Comment:
I've removed the canceller, but kept the explicit Context - it just feels to me like in most uses of this, it helps to have the context be explicit.  I can switch to `Context::current()` if you prefer.

Reply:
I do think it's worth removing the `canceller` argument from `activeWalk()` - we are relying on it being the same as `context->canceller()` (so that any computes see the same canceller as we're checking in `activeWalk()` itself), so having it as an extra argument just gives us a chance to get it wrong.